protocols/relay/src/v2/client/transport.rs

```diff
@@ -0,0 +1,509 @@
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::v2::client::RelayedConnection;
+use crate::v2::RequestId;
+use futures::channel::mpsc;
+use futures::channel::oneshot;
+use futures::future::{BoxFuture, Future, FutureExt};
+use futures::sink::SinkExt;
+use futures::stream::{Stream, StreamExt};
+use libp2p_core::either::{EitherError, EitherFuture, EitherOutput};
+use libp2p_core::multiaddr::{Multiaddr, Protocol};
+use libp2p_core::transport::{ListenerEvent, TransportError};
+use libp2p_core::{PeerId, Transport};
+use pin_project::pin_project;
+use std::collections::VecDeque;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use thiserror::Error;
+
+/// A [`Transport`] wrapping another [`Transport`] enabling client relay capabilities.
+///
+/// Allows the local node to:
+///
+/// 1. Use inner wrapped transport as before.
+///
+///    ```
+///    # use libp2p_core::{Multiaddr, multiaddr::{Protocol}, Transport, PeerId};
+///    # use libp2p_core::transport::memory::MemoryTransport;
+///    # use libp2p_relay::v2::client;
+///    # let inner_transport = MemoryTransport::default();
+///    # let (transport, behaviour) = client::Client::new_transport_and_behaviour(
+///    #     PeerId::random(),
+///    #     inner_transport,
+///    # );
+///    transport.dial(Multiaddr::empty().with(Protocol::Memory(42)));
+///    ```
+///
+/// 2. Establish relayed connections by dialing `/p2p-circuit` addresses.
+///
+///    ```
+///    # use libp2p_core::{Multiaddr, multiaddr::{Protocol}, Transport, PeerId};
+///    # use libp2p_core::transport::memory::MemoryTransport;
+///    # use libp2p_relay::v2::client;
+///    # let inner_transport = MemoryTransport::default();
+///    # let (transport, behaviour) = client::Client::new_transport_and_behaviour(
+///    #     PeerId::random(),
+///    #     inner_transport,
+///    # );
+///    let dst_addr_via_relay = Multiaddr::empty()
+///        .with(Protocol::Memory(40)) // Relay address.
+///        .with(Protocol::P2p(PeerId::random().into())) // Relay peer id.
+///        .with(Protocol::P2pCircuit) // Signal to connect via relay and not directly.
+///        .with(Protocol::Memory(42)) // Destination address.
+///        .with(Protocol::P2p(PeerId::random().into())); // Destination peer id.
```

Comment:
```suggestion
///    # let relay_id = PeerId::random();
///    # let destination_id = PeerId::random();
///    let dst_addr_via_relay = Multiaddr::empty()
///        .with(Protocol::Memory(40)) // Relay address.
///        .with(Protocol::P2p(relay_id.into())) // Relay peer id.
///        .with(Protocol::P2pCircuit) // Signal to connect via relay and not directly.
///        .with(Protocol::P2p(destination_id.into())); // Destination peer id.
```
I would not add a destination address here:  if I understand the specs correctly, the destination address is only used if active dialing is enabled, but this functionality is considered deprecated and therefore imo should not be used in an example

> Note that active relay functionality is considered deprecated for security reasons, at least in public relays

---

A general question about this: how can active dialing be even possible in relay v2 in the first place? 
If I understand it correctly, active dialing is needed when the destionation is not connected.
But the specs state that:
1. The relay only relays a connection to the destination if there is a reservation slot:
> If the relay has a reservation (..) from the peer, then it opens the second hop of the connection using the stop protocol;
2. When a destination disconnects, they loose their reservation:
> Note that the reservation remains valid until its expiration, as long as there is an active connection from the peer to the relay. If the peer disconnects, the reservation is no longer valid.

So there is no situation possible in which a destination peer has a reservation, but needs active dialing because it is not connected. Am I missing something here?

Reply:
Thanks for catching that. Indeed the destination address is not needed here and is misleading. Suggestion applied via 5d9f95af7b22c8cf3b2d09b295bde326f51d57ca.

---

> A general question about this: how can active dialing be even possible in relay v2 in the first place?

An active relay would have to open a connection even if it does not have a reservation for the destination. I don't think this is well covered in the specification today, as you noted above. It would probably need to be along the lines of:

> If the relay has a reservation (..) from the peer, then it opens the second hop of the connection using the stop protocol; **In case it is in active-relay mode, it would establish a connection to the destination first in case none yet exists**.

We haven't invested much time into active relaying on Circuit Relay v2, though, unless I am missing something, we can do so in the future without any breaking changes on the protocol layer.